code/datums/components/storage.dm

```diff
@@ -0,0 +1,762 @@
+#define COLLECT_ONE 0
+#define COLLECT_EVERYTHING 1
+#define COLLECT_SAME 2
+
+#define DROP_NOTHING 0
+#define DROP_AT_PARENT 1
+#define DROP_AT_LOCATION 2
+
+// External storage-related logic:
+// /mob/proc/ClickOn() in /_onclick/click.dm - clicking items in storages
+// /mob/living/Move() in /modules/mob/living/living.dm - hiding storage boxes on mob movement
+
+/datum/component/storage
+	var/list/can_hold								//if this is set, only things in this typecache will fit.
+	var/list/cant_hold								//if this is set, anything in this typecache will not be able to fit.
+
+	var/list/mob/is_using = list()							//lazy list of mobs looking at the contents of this storage.
+
+	var/locked = FALSE								//when locked nothing can see inside or use it.
+
+	var/max_w_class = WEIGHT_CLASS_SMALL			//max size of objects that will fit.
+	var/max_combined_w_class = 14					//max combined sizes of objects that will fit.
+	var/max_items = 7								//max number of objects that will fit.
+
+	var/emp_shielded = FALSE
+
+	var/silent = FALSE								//whether this makes a message when things are put in.
+	var/click_gather = FALSE						//whether this can be clicked on items to pick it up rather than the other way around.
+	var/rustle_sound = TRUE							//play rustle sound on interact.
+	var/allow_quick_empty = FALSE					//allow empty verb which allows dumping on the floor of everything inside quickly.
+	var/allow_quick_gather = FALSE					//allow toggle mob verb which toggles collecting all items from a tile.
+
+	var/collection_mode = COLLECT_EVERYTHING
+
+	var/insert_preposition = "in"					//you put things "in" a bag, but "on" a tray.
+
+	var/display_numerical_stacking = FALSE			//stack things of the same type and show as a single object with a number.
+
+	var/obj/screen/storage/boxes					//storage display object
+	var/obj/screen/close/closer						//close button object
+
+	var/allow_big_nesting = FALSE					//allow storage objects of the same or greater size.
+
+	var/atom/real_location						//where anything stored actually is.

```

Comment:
making a note here that cyberboss told me to remove this and instead make remote storages use another component.